imbi/endpoints/components.py

```diff
@@ -0,0 +1,177 @@
+from __future__ import annotations
+
+import enum
+import re
+
+import pydantic
+import typing_extensions as typing
+
+from imbi import errors
+from imbi.endpoints import base
+
+
+class ComponentStatus(str, enum.Enum):
+    ACTIVE = 'Active'
+    DEPRECATED = 'Deprecated'
+    FORBIDDEN = 'Forbidden'
+
+
+class Component(pydantic.BaseModel):
+    package_url: str = pydantic.constr(pattern=r'^pkg:')
+    name: str
+    status: ComponentStatus
+    icon_class: str
+    active_version: typing.Union[str, None]
+    home_page: typing.Union[str, None]
+
+
+class ComponentToken(base.PaginationToken):
+    """Pagination token that includes the starting package URL"""
+    def __init__(self, *, starting_package: str = '', **kwargs) -> None:
+        super().__init__(starting_package=starting_package, **kwargs)
+
+    def with_first(self, value: dict[str, object]) -> typing.Self:
+        kwargs = self.as_dict(starting_package=value['package_url'])
+        return ComponentToken(**kwargs)
+
+
+class CollectionRequestHandler(base.PaginatedCollectionHandler):
+    NAME = 'components'
+    ITEM_NAME = 'component'
+    ID_KEY = 'package_url'
+    FIELDS = [
+        'package_url', 'name', 'status', 'home_page', 'icon_class',
+        'active_version'
+    ]
+    DEFAULTS = {
+        'status': 'Active',
+        'icon_class': 'fas fa-save',
+        'active_version': None
+    }
+
+    COLLECTION_SQL = re.sub(
+        r'\s+', ' ', """\
+        SELECT package_url, "name", status, home_page, icon_class,
+               active_version, created_at, created_by,
+               last_modified_at, last_modified_by
+          FROM v1.components
+         WHERE package_url > %(starting_package)s
+         ORDER BY package_url ASC
+         LIMIT %(limit)s
+        """)
+    GET_SQL = re.sub(
+        r'\s+', ' ', """\
+        SELECT package_url, "name", status, home_page, icon_class,
+               active_version, created_at, created_by,
+               last_modified_at, last_modified_by
+          FROM v1.components
+         WHERE package_url = %(package_url)s
+        """)
```

Comment:
Is this used? Isn't only `COLLECTION_SQL` used for the GET?

Reply:
Nah ... there is an implicit usage of `GET_SQL` at the end of `CRUDRequestHandler._post` when it calls `self._get()`